### PR TITLE
Integrate with GitHub issues

### PR DESCRIPTION
**Pull request by jerivas**
## Link to Trello card (if applicable)
https://trello.com/c/ApEH2ks0/222-i-want-to-see-github-issues-in-metecho-that-i-can-work-on

## Description
_The commit messages say what you did; this should explain why and/or how._
- [x] Description is in Trello card

## Steps to test/reproduce

1. Visit `/api/issues/` to browse all available issues
2. You can filter issues by querystring, the filters are:
    - `project`: (hash ID) limit issues to a single Metecho project
    - `is_attached`: (boolean) filter issues by whether or not they are attached to an Epic or Task
    - `search`: (text) search in issue titles and numbers
3. To update the issues for a given project: `POST /api/project/:id/refresh_github_issues/`
4. Task and Epics now have an `issue` field in their payload. Set it to an issue hash ID to attach, or `null` to dettach.
5. Projects have a new `currently_fetching_issues` boolean field to indicate if a fetch is currently in progress. This can be used to disable the "Sync Issues" button to prevent multiple concurrent syncs

A new variable has been added to the JS globals so you can read it in the frontend: `GITHUB_ISSUE_LIMIT` . You can use this value in the string that says `showing the first ${GITHUB_ISSUE_LIMIT} issues`.

## Show me
_Provide screenshots/animated gifs/videos if necessary._

### CODE REVIEW

**Comment by jerivas** [comment]
@jgerigmeyer @oluoluoxenfree I still have to write tests, but right now the new API endpoints are in place:

1. Visit `/api/issues/` to browse all available issues
2. You can filter issues by querystring, the filters are:
    - `project`: (hash ID) limit issues to a single Metecho project
    - `is_attached`: (boolean) separate issues attached to an Epic or Task
    - `search`: (text) search in issue titles and numbers
3. To update the issues for a given project: `POST /api/project/:id/refresh_github_issues/`
4. Task and Epics now have an `issue` field in their payload. Set it to an issue hash ID to attach, or `null` to dettach.
5. Projects have a new `is_fetching_github_issues` boolean field to indicate if a fetch is currently in progress. This can be used to disable the "Sync" button to prevent multiple concurrent syncs

---

@oluoluoxenfree Our projects have a human-browsable API explorer you can use to try out these endpoints. Once you checkout this PR you can visit the URLs listed above in your browser and you'll get the response you expect.

For point 2, you can use the API explorer to apply filters

image

For point 4, you'll get an error about GET not being allowed (for security, actions like this require POST). You can click the POST button to actually run the action (the data you send doesn't matter). You will get an empty `202 Created` response immediately as the action is completed asynchronously. After a few seconds you can browse the Issues list and new content should appear there.

image

Finally, don't feel pressure to start working on this yet, I know you're busy with the other task 😊

**Comment by jgerigmeyer** [comment]
[I want to see GitHub issues in Metecho that I can work on.](https://trello.com/c/ApEH2ks0/222-i-want-to-see-github-issues-in-metecho-that-i-can-work-on)

File: metecho/api/models.py [hunk 1]
```diff
@@ -400,6 +402,24 @@ def finalize_populate_github_users(self, *, error=None, originating_user_id):
         else:
             self.notify_error(error, originating_user_id=originating_user_id)
 
+    def queue_refresh_github_issues(self, *, originating_user_id):
+        from .jobs import refresh_github_issues_job
+
+        if not self.currently_fetching_issues:
+            self.currently_fetching_issues = True
+            self.save()
```
Comment (jgerigmeyer) [inline metecho/api/models.py:410]: I think we also want a `self.notify_changed(originating_user_id=originating_user_id)` after this `save()`?

File: metecho/api/serializers.py [hunk 2]
```diff
@@ -48,6 +49,48 @@ def to_representation(self, value):
...
+            return None
+
+    def get_task(self, issue):
+        try:
+            return {
+                "id": str(issue.task.id),
+                "name": issue.task.name,
+                "status": issue.task.status,
+                "review_status": issue.task.review_status,
+            }
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:89]: Looking at the mockup, I see why you nested these here. To generate the link to the Task Detail page, though, I think we'll also need the `issue.task.epic.slug` and `issue.task.slug`. (Unless we just want the `id` and make the front end look up the necessary info.)

File: metecho/api/serializers.py [hunk 3]
```diff
@@ -48,6 +49,48 @@ def to_representation(self, value):
...
+        )
+        fields = read_only_fields
+
+    def get_epic(self, issue):
+        try:
+            return {
+                "id": str(issue.epic.id),
+                "name": issue.epic.name,
+                "status": issue.epic.status,
+            }
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:78]: To generate the link to the Epic Detail page, I think we'll also need the `issue.epic.slug`.

File: metecho/api/serializers.py [hunk 4]
```diff
@@ -350,6 +398,7 @@ class Meta:
             "origin_sha",
             "branch_diff_url",
             "pr_url",
+            "issue",
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:401]: It's not ideal, but I think all of these `serializer` changes should make their way into the relevant `docs/` files too.

**Review by jgerigmeyer — COMMENTED** [review]
@jerivas Looks like a good start!

@oluoluoxenfree When you're ready to start work on this, let's schedule a mtg to talk through the front end requirements here. 👍

File: metecho/api/serializers.py (hunk 2 above)
Comment (jerivas) [reply metecho/api/serializers.py:89]: Would you say nesting is better in this case or we should just return the epic/task ID?

File: metecho/api/models.py (hunk 1 above)
Comment (jerivas) [reply metecho/api/models.py:410]: Fixed in 94e344587db6a311467b76c6be3f1831c8a2c96f

File: metecho/api/serializers.py (hunk 2 above)
Comment (jerivas) [reply metecho/api/serializers.py:89]: For now fixed in 94e344587db6a311467b76c6be3f1831c8a2c96f

File: metecho/api/serializers.py (hunk 4 above)
Comment (jerivas) [reply metecho/api/serializers.py:404]: Fixed in 30c74071cb041faa423eac9b35bbb066d05a4a5f. Also went back and documented previous changes.

File: metecho/api/serializers.py (hunk 3 above)
Comment (jerivas) [reply metecho/api/serializers.py]: Fixed in 94e344587db6a311467b76c6be3f1831c8a2c96f

File: metecho/api/serializers.py (hunk 2 above)
Comment (jgerigmeyer) [reply metecho/api/serializers.py]: I like this for now -- unless we refactor, we aren't guaranteed to have the full list of Tasks in the front end store at this point. 👍

File: metecho/context_processors.py [hunk 5]
```diff
@@ -10,6 +10,7 @@ def env(request):
     )
     GLOBALS = {
         "SITE": site_profile.data,
+        "GITHUB_ISSUE_LIMIT": settings.GITHUB_ISSUE_LIMIT,
```
Comment (jerivas) [inline metecho/context_processors.py:13]: @jgerigmeyer @oluoluoxenfree I exposed this setting to the JS globals so you can read it in the frontend. You can use this value in the string that says `showing the first ${GITHUB_ISSUE_LIMIT} issues`.

**Comment by jerivas** [comment]
@jgerigmeyer tests are now complete 🚀

**Review by jgerigmeyer — COMMENTED** [review]
@jerivas Looks good!

As far as populating the list initially, I'm thinking the front end can automatically trigger a "resync" when the modal is opened, if the list is empty. Does that seem sensible? (I think that's what we're doing with GH collaborators...)

**Comment by jerivas** [comment]
> the front end can automatically trigger a "resync" when the modal is opened

Sounds good to me 👍

File: metecho/api/serializers.py [hunk 6]
```diff
@@ -229,6 +302,12 @@ class EpicSerializer(HashIdModelSerializer):
     branch_url = serializers.SerializerMethodField()
     branch_diff_url = serializers.SerializerMethodField()
     pr_url = serializers.SerializerMethodField()
+    issue = serializers.PrimaryKeyRelatedField(
+        queryset=GitHubIssue.objects.all(),
+        pk_field=serializers.CharField(),
+        required=False,
+        allow_null=True,
+    )
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:310]: @jerivas For issues on both Epics and Tasks, I'm wondering what you think about doing a one-way (readonly) nested serialized representation of an attached issue instead of a simple foreign key. That would avoid the front end having to either fetch the full list of all attached issues for an entire project on page-load, or doing lots of individual requests to fetch the attached issue for specific tasks/epics while navigating the site.

Basically, I'm thinking that the front end might store the lists of available and attached tasks in-memory only while the modal is open (not in the Redux store). But we need to display attached issues throughout the site, so as it's written now we'd probably need to fetch all attached issues for a project when we load the project, even if there are multiple pages.

We can talk through this tomorrow if that'd be useful.

Comment (jerivas) [reply metecho/api/serializers.py:310]: I think that should work, yes. It's similar to what we did with `TaskSerializer.epic` right?

That being said, in Top Dog we favored filtered lists endpoints instead of nested fields because they play better with DRL. I think it might be worth to move towards that in preparation to the upcoming DRL story. What do you think?

Comment (jgerigmeyer) [reply metecho/api/serializers.py:310]: Ok, that makes sense -- we can keep them as separate lists.

Maybe I won't worry about fetching the entire list of attached issues and storing it in the redux store. Instead we can fetch on-demand for a particular epic or task if there's one attached, and then only fetch the full per-project list when the modal is opened.

**Review by jgerigmeyer — COMMENTED** [review]
@oluoluoxenfree This is coming along nicely! 🎉 I backfilled the last bit of test coverage, and did some review/cleanup (see: https://github.com/oddbird/Metecho/pull/1232/files/1689849489dd319fe074191e7d9f8ec704a2fba3..b604271939b56ab7a36722ffbf6b6eba486d8efb).

This is in a good place, but I don't think it's ready to merge to `main` as-is. If you prefer, you can branch off this PR for the next steps, to keep PR diffs manageable.

Next steps:

- Show attached issues on Epic and Task pages.
- Allow attaching an issue to an exiting Epic or Task.
- Allow refreshing the list of issues.
- Allow searching the list of issues.
- Add pagination to the lists of issues.

That's a lot, and it doesn't need to happen all at once. I'd recommend doing them roughly in the order listed above, but I'm open to your preference on that.

Feel free to put something on my calendar for tomorrow to talk through next steps in detail, if that'd be useful.

File: src/js/components/githubIssues/selectIssueModal.tsx [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+                  {attachedIssues.length ? (
+                    attachedIssues.map((issue, idx) => (
+                      <div key={idx} className="slds-p-vertical_x-small">
+                        <Radio
+                          labels={{ label: `#${issue.number}: ${issue.title}` }}
+                          name="attached-github-issue"
+                          checked
+                          value={issue.id}
+                          disabled
+                        />
```
Comment (jgerigmeyer) [inline src/js/components/githubIssues/selectIssueModal.tsx:166]: @oluoluoxenfree I also notice that this is missing some items from the mockup. Once an issue is attached, it looks like we want to display the name of the task/epic that it's attached to, with a link to the task/epic, and the status of the task/epic.

https://xd.adobe.com/view/bee12238-9700-471f-b104-3aa900d43f6e-ebb2/screen/98ab3852-0936-4ce1-a77a-89ba5f670d6e/

**Comment by stacyk** [comment]
I added a tiny bit of style to the buttons, but I was not able to see a list of issues. I went to the [local address/](http://localhost:8080/api/issues/) as per the directions, but wasn't sure what to do there as it says count: 0 even after filtering for Metecho Test. I see there are 4 open issues in GH for that project. I assume I am missing some other bit of instruction to make this work locally?

I know there is a part 2 branch and I was thinking about doing style/clean up here first before I do anything in part 2 (if needed).

**Comment by oluoluoxenfree** [comment]
@stacyk I think you have to POST an empty request to that endpoint?

**Comment by stacyk** [comment]
> @stacyk I think you have to POST an empty request to that endpoint?

@oluoluoxenfree Is that something I do in the browser? I don't really know where to get started on this.

**Comment by jgerigmeyer** [comment]
> I don't really know where to get started on this.

@stacyk To populate the list of Issues for a specific Metecho Project:

1. Find the `id` of the relevant Project (simplest is to find it in the list at http://localhost:8080/api/projects/ and copy the id)
2. Go to http://localhost:8080/api/projects/<id>/refresh_github_issues/ (replace `<id>` with the copied id)
3. Click `POST` btn at the bottom of the page

After a few seconds, the issues should be updated for that Project, and should appear in the app.

File: metecho/api/serializers.py [hunk 8]
```diff
@@ -88,6 +89,80 @@ def get_choices(self, cutoff=None):  # pragma: nocover
...
+        {
+            "properties": {
+                "id": {"type": "string", "format": "HashID"},
+                "name": {"type": "string"},
+                "status": {"type": "string"},
+                "review_status": {"type": "string"},
+                "review_valid": {"type": "boolean"},
+                "pr_is_open": {"type": "boolean"},
+                "slug": {"type": "string"},
+                "epic_slug": {"type": "string", "nullable": True},
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:145]: @jerivas I wasn't 100% sure if this is right way to indicate that this could be `None`?

File: metecho/api/serializers.py [hunk 9]
```diff
@@ -88,6 +89,80 @@ def get_choices(self, cutoff=None):  # pragma: nocover
...
+        try:
+            return {
+                "id": str(issue.task.id),
+                "name": issue.task.name,
+                "status": issue.task.status,
+                "review_status": issue.task.review_status,
+                "review_valid": issue.task.review_valid,
+                "pr_is_open": issue.task.pr_is_open,
+                "slug": issue.task.slug,
+                "epic_slug": issue.task.epic.slug if issue.task.epic else None,
```
Comment (jgerigmeyer) [inline metecho/api/serializers.py:160]: @jerivas This was breaking for tasks that don't have an epic, so I added this `if`.

**Review by jgerigmeyer — APPROVED** [review]
Looks good!

File: metecho/api/serializers.py (hunk 8 above)
Comment (jerivas) [reply metecho/api/serializers.py:145]: Makes sense to me 👍

File: .devcontainer/docker-compose.dev.yml [hunk 10]
```diff
@@ -2,6 +2,7 @@ version: '3'
 
 services:
   web:
+    platform: linux/amd64
```
Comment (jgerigmeyer) [inline .devcontainer/docker-compose.dev.yml:5]: I'm not sure what practical difference this is supposed to make? I've been building and running locally without it (on M1) without any trouble... 😕

Comment (jgerigmeyer) [reply .devcontainer/docker-compose.dev.yml:5]: @stacyk Did you hit a particular issue with the Metecho build that required adding this? Reading the Docker docs, it sounds like this setting should only be used as a last resort. https://docs.docker.com/desktop/mac/apple-silicon/#known-issues

Comment (stacyk) [reply .devcontainer/docker-compose.dev.yml:5]: @jgerigmeyer I think this was the first of three things I did to make it run. I didn't go back and remove it to see if the other two were fine on their own. We could certainly try and take it away and if I have a problem moving forward, I will be sure to let you know ;)

Comment (jgerigmeyer) [reply .devcontainer/docker-compose.dev.yml:5]: Ok, my theory is that it might be needed on MetaDeploy because of the additional install_sfdx script? I'm going to try removing it here...

**Comment by jgerigmeyer** [comment]
Closing this is lieu of #1707.
